### ISSUE TITLE
[tvOS] Cannot create databases in default CBLManager

### ISSUE THREAD

**Issue by ashvindersingh** [couchbase/couchbase-lite-ios#1081]
Steps to reproduce: 
- Git clone https://github.com/couchbaselabs/iOS-WorkerBee.git
- Copy and link 1.2.0-101 tvOS binaries of Couchbselite.framework and CouchbaseLiteListener.framework
- Select the tvOS scheme and simulator
- Add the following option in scheme in section "Run" with params "-RunTest Test1_CreateDocs"
- Run 
- Get the following exception: https://gist.github.com/ashvindersingh/215fcb8579592fc7fa11#file-tvos-workerbee-logs-L131

**Comment by pasin** [comment]
From @ashvindersingh's log:

```
03:29:50.289| WARNING: Upgrade skipped: Database upgrading class is not present.
2016-01-29 03:29:50.288 headlessbee-Appletv[47876:8486033] *** Assertion failure in +[SavedTestRun modelForNewDocumentInDatabase:], /Users/jenkins/jenkins/workspace/couchbase-lite-ios-builds/couchbase-lite-tvos/Source/API/CBLModel.m:62
2016-01-29 03:29:50.297 headlessbee-Appletv[47876:8486033] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: database'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010eccfff5 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x000000010e74adeb objc_exception_throw + 48
	2   CoreFoundation                      0x000000010eccfe5a +[NSException raise:format:arguments:] + 106
	3   Foundation                          0x000000010e09662e -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 198
	4   headlessbee-Appletv                 0x000000010c2ca9fc +[CBLModel modelForNewDocumentInDatabase:] + 208
	5   headlessbee-Appletv                 0x000000010c29afe1 +[SavedTestRun forTest:] + 113
	6   headlessbee-Appletv                 0x000000010c29838a -[BeeTest tearDown] + 202
	7   headlessbee-Appletv                 0x000000010c29aa78 -[BeeCouchTest tearDown] + 136
	8   headlessbee-Appletv                 0x000000010c2981db -[BeeTest setRunning:] + 731
	9   headlessbee-Appletv                 0x000000010c29de3d -[BeeCouchMultipleTest runMultiple] + 8045
	10  headlessbee-Appletv                 0x000000010c29ba99 -[BeeCouchMultipleTest setUp] + 73
	11  headlessbee-Appletv                 0x000000010c2a5536 -[Test1_CreateDocs setUp] + 54
	12  headlessbee-Appletv                 0x000000010c2981c3 -[BeeTest setRunning:] + 707
	13  headlessbee-Appletv                 0x000000010c297382 -[HeadlessBeeAppDelegate runTest:] + 146
	14  headlessbee-Appletv                 0x000000010c297028 -[HeadlessBeeAppDelegate runTests] + 520
	15  Foundation                          0x000000010e032cd9 __NSFireDelayedPerform + 402
	16  CoreFoundation                      0x000000010ec2fe14 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
	17  CoreFoundation                      0x000000010ec2f9c1 __CFRunLoopDoTimer + 1089
	18  CoreFoundation                      0x000000010ebf13d1 __CFRunLoopRun + 1937
	19  CoreFoundation                      0x000000010ebf09b8 CFRunLoopRunSpecific + 488
	20  GraphicsServices                    0x00000001105c9ad2 GSEventRunModal + 161
	21  UIKit                               0x000000010caaf8fc UIApplicationMain + 171
	22  headlessbee-Appletv                 0x000000010c296bef main + 111
	23  libdyld.dylib                       0x000000010f1b29e9 start + 1
)
```

**Comment by pasin** [comment]
I was able to reproduce the issue when I don't have `-ObjC` set in `Other Linker Flags`. Could you please add `-ObjC` to the target build settings?

**Comment by snej** [comment]
Also, the assertion failure is due to a lack of error checking in Worker Bee itself. `+[SavedTestRun forTest:]` doesn't check whether it got a nil database, and then passes that nil pointer to CBLModel.

**Comment by ashvindersingh** [comment]
On the simulator the workerbee app runs ok but getting this error when run on the appletv device:

```
2016-01-18 08:39:39.115 HeadlessBee-AppletV[180:10736] ERROR: Can't write to log: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"
2016-01-18 08:39:39.115 HeadlessBee-AppletV[180:10736] logMessage: 
100000 docs; 0.00%; 0.00%; 0.00%; 0.00%; 0.00%; 0.00%; 0.00%; ;
2016-01-18 08:39:39.130 HeadlessBee-AppletV[180:10736] ERROR: Can't write to log: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"
2016-01-18 08:39:39.130 HeadlessBee-AppletV[180:10736] ERROR: Can't write to log: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"
2016-01-18 08:39:39.130 HeadlessBee-AppletV[180:10736] logMessage: (Elapsed time: 0.709 sec)
2016-01-18 08:39:39.131 HeadlessBee-AppletV[180:10736] *** Assertion failure in +[SavedTestRun modelForNewDocumentInDatabase:], /Users/jenkins/jenkins/workspace/couchbase-lite-ios-builds/couchbase-lite-tvos/Source/API/CBLModel.m:62
```

**Comment by pasin** [comment]
@ashvindersingh Can you run the test with couchbase-lite-ios instead of using worker bee? The replication test will fail as invalid SSL certificate but at least, you should be able to verify if the binary works or not.

**Comment by snej** [comment]
It's saying it can't write to its log file.
On tvOS the only location that's writeable by an app is its Caches directory.
So the location of the log file would need to be changed to that directory … if you even need the logs to be saved to disk at all. Don't you just copy them from the Xcode console instead?

**Comment by ashvindersingh** [comment]
Yes, true, so I commented the function in
BeeTest.m to be like this:
```
- (void) writeToOutput: (NSString*)message {
    /*
    NSAssert(_output, @"Output isn't open");
    NSData* data = [message dataUsingEncoding: NSUTF8StringEncoding];
    NSInteger written = [_output write: data.bytes maxLength: data.length];
    if (written < 0)
        NSLog(@"ERROR: Can't write to log: %@", _output.streamError);
    else
        NSAssert(written == data.length, @"Only wrote %ld bytes of %lu", (long)written, (unsigned long)data.length);
    [_output write: (const uint8_t*)"\n" maxLength: 1];
     */
}

```
but still getting the following error:

```
2016-02-01 22:27:53.950 HeadlessBee-AppletV[190:6181] *** Assertion failure in +[SavedTestRun modelForNewDocumentInDatabase:], /Users/jenkins/jenkins/workspace/couchbase-lite-ios-builds/couchbase-lite-tvos/Source/API/CBLModel.m:62
```

The problematic line is in file SavedTestRun.m, line# 55, where the assertion comes.
```
  SavedTestRun* instance = [SavedTestRun modelForNewDocumentInDatabase:[self database]];
```

**Comment by snej** [comment]
Yes, [I answered that earlier](https://github.com/couchbase/couchbase-lite-ios/issues/1081#issuecomment-176928949).

**Comment by snej** [comment]
* The assertion failure says that in `modelForNewDocumentInDatabase`, the `database` parameter is nil.
* So that means that in SavedTestRun, `[self database]` returns nil.
* Look at the `-database` method: it calls `[[CBLManager sharedInstance] databaseNamed: @"workerbee-tests" error: NULL]` but doesn't handle errors (nil results). It also doesn't pass an NSError to the call so it won't know what the specific error was.
* If you fix that then you can find out what the actual error is, and at least die with a relevant error instead of the assertion failure that happens later.

**Comment by pasin** [comment]
The issue is that [CBLManager sharedInstance] returns nil as the default directory path (Application Support) doesn't allow in tvOS.

```
Error Domain=NSCocoaErrorDomain Code=513 "You don’t have permission to save the file “CouchbaseLite” in the folder “Application Support”." UserInfo={NSFilePath=/var/mobile/Containers/Data/Application/AE3AF6AE-69EF-405E-B1DA-0E315A8EB4BF/Library/Application Support/CouchbaseLite, NSUnderlyingError=0x145d1f160 {Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"}}
```

**Comment by snej** [comment]
But the shared CBLManager on tvOS uses the Caches directory instead of App Support, for exactly that reason.

**Comment by snej** [comment]
Oh fuck — the commit that enabled that (on the special 1.1.1-appletv branch) didn't get merged into 1.2. Need to get that in.

**Comment by snej** [comment]
Fixed with #1089, commit 6297441fb74e512ef27c05ba1b72a745b6ed5007 (forgot to tag this bug in the commit)
